File: simpledroidble/simpledroidble/src/main/java/org/simpleble/android/Peripheral.kt
package org.simpleble.android

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class Peripheral internal constructor(newAdapterId: Long, newInstanceId: Long) {

    private var instanceId: Long = newInstanceId
    private var adapterId: Long = newAdapterId

    private val _onConnected = MutableSharedFlow<Unit>()
    private val _onConnectionActive = MutableSharedFlow<Boolean>()
    private val _onDisconnected = MutableSharedFlow<Unit>()

    private val callbacks = object : Callback {
        override fun onConnected() {
            CoroutineScope(Dispatchers.Main).launch {
                _onConnected.emit(Unit)
            }
            CoroutineScope(Dispatchers.Main).launch {
                _onConnectionActive.emit(true)
            }
        }

        override fun onDisconnected() {
            CoroutineScope(Dispatchers.Main).launch {
                _onDisconnected.emit(Unit)
            }
            CoroutineScope(Dispatchers.Main).launch {
                _onConnectionActive.emit(false)
            }
        }
    }

    init {
        Log.d("SimpleBLE", "Peripheral ${this.hashCode()}.init")
        nativePeripheralRegister(adapterId, instanceId, callbacks)
    }

    val identifier: String get() {
        return nativePeripheralIdentifier(adapterId, instanceId) ?: ""
    }

    val address: BluetoothAddress get() {
        return BluetoothAddress(nativePeripheralAddress(adapterId, instanceId) ?: "")
    }

    val addressType: BluetoothAddressType get() {
        return BluetoothAddressType.fromInt(nativePeripheralAddressType(adapterId, instanceId))
    }

    val rssi: Int get() {
        return nativePeripheralRssi(adapterId, instanceId)
    }

    val txPower: Int get() {
        return nativePeripheralTxPower(adapterId, instanceId)
    }

    val mtu: Int get() {
        return nativePeripheralMtu(adapterId, instanceId)
    }

    suspend fun connect() {
        withContext(Dispatchers.IO) {
            nativePeripheralConnect(adapterId, instanceId)
        }
    }

    suspend fun disconnect() {
        withContext(Dispatchers.IO) {
            nativePeripheralDisconnect(adapterId, instanceId)
        }
    }

    val isConnected: Boolean get() {
        return false
    }

    val isConnectable: Boolean get() {
        return false
    }

    val isPaired: Boolean get() {
        return false
    }

    fun unpair() {
    }

    fun services(): List<Service> {
        return emptyList()
    }

    fun manufacturer_data(): Map<Int, ByteArray> {
        return emptyMap()
    }

    fun read(service: BluetoothUUID, characteristic: BluetoothUUID): ByteArray {
        return ByteArray(0)
    }

    fun write_request(service: BluetoothUUID, characteristic: BluetoothUUID, data: ByteArray) {
    }

    fun write_command(service: BluetoothUUID, characteristic: BluetoothUUID, data: ByteArray) {
    }

    fun notify(
        service: BluetoothUUID,
        characteristic: BluetoothUUID
    ): MutableSharedFlow<ByteArray> {
        val payloadFlow = MutableSharedFlow<ByteArray>()
        val dataCallback = object : DataCallback {
            override fun onDataReceived(data: ByteArray) {
                CoroutineScope(Dispatchers.Main).launch {
                    payloadFlow.emit(data)
                }
            }
        }

        nativePeripheralNotify(adapterId, instanceId, service.toString(), characteristic.toString(), dataCallback)
        return payloadFlow
    }

    fun indicate(
        service: BluetoothUUID,
        characteristic: BluetoothUUID
    ): MutableSharedFlow<ByteArray> {
        val payloadFlow = MutableSharedFlow<ByteArray>()
        val dataCallback = object : DataCallback {
            override fun onDataReceived(data: ByteArray) {
                CoroutineScope(Dispatchers.Main).launch {
                    payloadFlow.emit(data)
                }
            }
        }

        nativePeripheralIndicate(adapterId, instanceId, service.toString(), characteristic.toString(), dataCallback)
        return payloadFlow
    }

    fun unsubscribe(service: BluetoothUUID, characteristic: BluetoothUUID) {
        nativePeripheralUnsubscribe(adapterId, instanceId, service.toString(), characteristic.toString())
    }

    fun read(
        service: BluetoothUUID,
        characteristic: BluetoothUUID,
        descriptor: BluetoothUUID
    ): ByteArray {
        return ByteArray(0)
    }

    fun write(
        service: BluetoothUUID,
        characteristic: BluetoothUUID,
        descriptor: BluetoothUUID,
        data: ByteArray
    ) {
    }

    val onConnected get() = _onConnected

    val onDisconnected get() = _onDisconnected

    val onConnectionActive get() = _onConnectionActive

    /// ----------------------------------------------------------------------------

    private external fun nativePeripheralRegister(adapterId: Long, instanceId: Long, callback: Callback)

    private external fun nativePeripheralIdentifier(adapterId: Long, instanceId: Long): String?

    private external fun nativePeripheralAddress(adapterId: Long, instanceId: Long): String?

    private external fun nativePeripheralAddressType(adapterId: Long, instanceId: Long): Int

    private external fun nativePeripheralRssi(adapterId: Long, instanceId: Long): Int

    private external fun nativePeripheralTxPower(adapterId: Long, instanceId: Long): Int

    private external fun nativePeripheralMtu(adapterId: Long, instanceId: Long): Int

    private external fun nativePeripheralConnect(adapterId: Long, instanceId: Long)

    private external fun nativePeripheralDisconnect(adapterId: Long, instanceId: Long)

    private external fun nativePeripheralNotify(
        adapterId: Long,
        instanceId: Long,
        service: String,
        characteristic: String,
        dataCallback: DataCallback
    )

    private external fun nativePeripheralIndicate(
        adapterId: Long,
        instanceId: Long,
        service: String,
        characteristic: String,
        dataCallback: DataCallback
    )

    private external fun nativePeripheralUnsubscribe(
        adapterId: Long,
        instanceId: Long,
        service: String,
        characteristic: String)

    // ----------------------------------------------------------------------------

    private interface DataCallback {
        fun onDataReceived(data: ByteArray)
    }


    private interface Callback {
        fun onConnected()
        fun onDisconnected()
    }

}

File: simpleble/src/frontends/base/Peripheral.cpp
#include <simpleble/Peripheral.h>

#include <simpleble/Exceptions.h>
#include "PeripheralBase.h"

using namespace SimpleBLE;

bool Peripheral::initialized() const { return internal_ != nullptr; }

void* Peripheral::underlying() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().underlying();
}

std::string Peripheral::identifier() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().identifier();
}

SimpleBLE::BluetoothAddress Peripheral::address() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().address();
}

SimpleBLE::BluetoothAddressType Peripheral::address_type() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().address_type();
}

int16_t Peripheral::rssi() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().rssi();
}

int16_t Peripheral::tx_power() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().tx_power();
}

uint16_t Peripheral::mtu() const {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().mtu();
}

void Peripheral::connect() {
    if (!initialized()) throw Exception::NotInitialized();

    internal().connect();
}

void Peripheral::disconnect() {
    if (!initialized()) throw Exception::NotInitialized();

    internal().disconnect();
}

bool Peripheral::is_connected() {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().is_connected();
}

bool Peripheral::is_connectable() {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().is_connectable();
}

bool Peripheral::is_paired() {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().is_paired();
}

void Peripheral::unpair() {
    if (!initialized()) throw Exception::NotInitialized();

    internal().unpair();
}

std::vector<Service> Peripheral::services() {
    if (!initialized()) throw Exception::NotInitialized();

    if (is_connected()) {
        return internal().services();
    } else {
        return internal().advertised_services();
    }

    return internal().services();
}

std::map<uint16_t, ByteArray> Peripheral::manufacturer_data() {
    if (!initialized()) throw Exception::NotInitialized();

    return internal().manufacturer_data();
}

ByteArray Peripheral::read(BluetoothUUID const& service, BluetoothUUID const& characteristic) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    return internal().read(service, characteristic);
}

void Peripheral::write_request(BluetoothUUID const& service, BluetoothUUID const& characteristic,
                               ByteArray const& data) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    internal().write_request(service, characteristic, data);
}

void Peripheral::write_command(BluetoothUUID const& service, BluetoothUUID const& characteristic,
                               ByteArray const& data) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    internal().write_command(service, characteristic, data);
}

void Peripheral::notify(BluetoothUUID const& service, BluetoothUUID const& characteristic,
                        std::function<void(ByteArray payload)> callback) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    internal().notify(service, characteristic, std::move(callback));
}

void Peripheral::indicate(BluetoothUUID const& service, BluetoothUUID const& characteristic,
                          std::function<void(ByteArray payload)> callback) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    internal().indicate(service, characteristic, std::move(callback));
}

void Peripheral::unsubscribe(BluetoothUUID const& service, BluetoothUUID const& characteristic) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    internal().unsubscribe(service, characteristic);
}

ByteArray Peripheral::read(BluetoothUUID const& service, BluetoothUUID const& characteristic,
                           BluetoothUUID const& descriptor) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    return internal().read(service, characteristic, descriptor);
}

void Peripheral::write(BluetoothUUID const& service, BluetoothUUID const& characteristic,
                       BluetoothUUID const& descriptor, ByteArray const& data) {
    if (!initialized()) throw Exception::NotInitialized();
    if (!is_connected()) throw Exception::NotConnected();

    internal().write(service, characteristic, descriptor, data);
}

void Peripheral::set_callback_on_connected(std::function<void()> on_connected) {
    if (!initialized()) throw Exception::NotInitialized();

    internal().set_callback_on_connected(std::move(on_connected));
}

void Peripheral::set_callback_on_disconnected(std::function<void()> on_disconnected) {
    if (!initialized()) throw Exception::NotInitialized();

    internal().set_callback_on_disconnected(std::move(on_disconnected));
}

PeripheralBase &Peripheral::internal() {
    return *internal_;
}

const PeripheralBase &Peripheral::internal() const {
    return *internal_;
}
